### DISCUSSION TITLE
recursive autodiscover_tasks

### DISCUSSION THREAD

**Discussion by IterableTrucks** [celery/celery#7179] [Ideas]
# Checklist

* [x] I have checked the [issues list](https://github.com/celery/celery/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22Issue+Type%3A+Feature+Request%22+)
  for similar or identical feature requests.
* [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?utf8=%E2%9C%93&q=is%3Apr+label%3A%22PR+Type%3A+Feature%22+)
  for existing proposed implementations of this feature.
* [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the if the same feature was already implemented in the
  master branch.
* [ ] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

#### Related Issues
* None

#### Possible Duplicates
* None

# Brief Summary

 The autodiscover_tasks method checks modules in _packages_, but modules in subdirectories will not be imported. It will be convenient to add a recursive option which enables searching modules in _packages_ and their subdirectories.

# Design
## Architectural Considerations

autodiscover_tasks(packages=None, related_name='tasks', force=False) will be changed to autodiscover_tasks(packages=None, related_name='tasks', force=False,recursive=False) for backward compatibility

## Diagrams

N/A

## Alternatives

None

**Comment by auvipy** [comment]
can you share a use-case for this?

**Comment by BradWells** [comment]
I have a use case for this. I am using Celery with Django and have an app 'rest' that is primarily a rest backend. Inside that are sub-packages named after the different django models I have ('accounts', 'goals', 'events', etc), so that I can keep my Django urls and views separate and modular.

A few of these sub-packages have recurring tasks associated with them for celery-beat, so I made a tasks.py in each sub-package (technically for me jobs.py with `related_name='jobs'`, but same thing). I figured celery through Django-fixups would pick them up since they are in my INSTALLED_APPS, but it would only pick up rest/tasks.py, not rest/goals/tasks.py.  This means I need to list out `celery_app.autodiscover_tasks(['rest.goals'])` instead of just `celery_app.autodiscover_tasks()`, which is one more thing to forget during maintenance. 

I would love to just add `recursive=True` instead of maintaining this list.
